### ISSUE TITLE
Many-to-many "through" table with multiple rows of identical foreign key pairs only returns one result

### ISSUE THREAD

**Issue by jas** [sequelize/sequelize#3493]
I'm using the `mysql` dialect with `sequelize@^2.0.5`.

My database is setup such that I have `users` and `items`. Users can mark items as "owned", and for this I have an `ownership` table. My scenario allows for users to mark the same item as owned multiple times while including some additional data — for this example, a note — about one of the ownerships.

Here's an example of the `ownership` table:

id  | userId  | itemId  | notes
----|---------|---------|-----------------------
1   | 50      | 941     | This is my green one.
2   | 50      | 941     | This is my blue one. 

My models are defined as follows:

```js
var User = sequelize.define('user', {
  username: DataTypes.STRING
});

var Item = sequelize.define('item', {
  title: DataTypes.STRING
});

var Ownership = sequelize.define('ownership', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true
  },
  notes: DataTypes.TEXT
});

Item.belongsToMany(User, { through: Ownership });
User.belongsToMany(Item, { through: Ownership });
```

I am explicitly defining the `id` primary key on the `Ownership` model because of this paragraph in the Sequelize documentation:

[Relations/Associations -> Belongs-To-Many Associations](http://docs.sequelizejs.com/en/latest/docs/associations/#belongs-to-many-associations)
> By default the code above will add ProjectId and UserId to the UserProjects table, and remove any previsouly defined primary key attribute - **the table will be uniquely identified by the combination of the keys of the two tables**, and there is no reason to have other PK columns. To enforce a primary key on the UserProjects model you can add it manually.

When I fetch a user, I would like both of of the `Items` for the `User` to be available in the model, however only one is. The SQL that is output in the console is correct and does in fact return two results, but I think something internally in Sequelize is removing the other — perhaps looking at the foreign keys to determine the uniqueness of the `Item`. Based on the paragraph in the documentation, I assumed that adding the `id` attribute to the `Ownership` model would fix this and treat each one as a uniquely.

```js
User.find({
  where: { id: 50 },
  include: [{ all: true, nested: true }]
}).then(function(user) {
  console.log(user.toJSON());
});
```

**Expected**:

```json
{
  "id": 50,
  "username": "john",
  "items": [{
    "id": 1,
    "title": "foobar",
    "ownership": {
      "id": 1,
      "userId": 50,
      "itemId": 941,
      "notes": "This is my green one."
    }
  }, {
    "id": 1,
    "title": "foobar",
    "ownership": {
      "id": 2,
      "userId": 50,
      "itemId": 941,
      "notes": "This is my blue one."
    }
  }]
}
```

**Actual:**

```json
{
  "id": 50,
  "username": "john",
  "items": [{
    "id": 1,
    "title": "foobar",
    "ownership": {
      "id": 2,
      "userId": 50,
      "itemId": 941,
      "notes": "This is my blue one."
    }
  }]
}
```

In the meantime, I am able to workaround this by using `user.getItems()`...

```js
User.find({
  where: { id: 50 },
  include: [{ all: true, nested: true }]
}).then(function(user) {
  user.getItems({
    include: [{ all: true, nested: true }]
  }).then(function(items) {
    var data = user.toJSON();
    data.items = items;
    console.log(data.toJSON());
  });
});
```

...but this results in an extra query that I don't believe is necessary since the original query is returning the correct rows and is just not being translated correctly to a Sequelize model.

I hope this all makes sense, and I very much appreciate any help or direction if I am doing something wrong.

**Comment by mickhansen** [comment]
We don't support duplicate entries for through tables currently.
You'll have to manually setup associations to and from ownership and manually include it for it to work currently.

**Comment by eikaramba** [comment]
is there any update on this? Unfortunately it is still the case and frankly quite a big drawback IMHO. Can you @mickhansen maybe at least point into the right direction whether this is theoretically possible from your point of view. then maybe i could have a look into the code to see what can be done about this.

The problem is that i really want to preserve the relation and just adding another hasMany relation would mean that i loose the context of the target Table. Or in this example above, i would like to still iterate over all items and THEN look at the ownerships.

**Comment by edencorbin** [comment]
What are some good alternative approaches to this, I have the same issue? My current thinking is to make a new table that (to follow the example) stores userId, itemId and note, set each to unique: true, then running inserts and queries directly on this table and not using any defined relationship.

**Comment by eikaramba** [comment]
I settled too with the option to actually do multiple queries. Not nice and i still think this is a big drawback. :(

**Comment by eikaramba** [comment]
Yes this is issue is very relevant. I honestly think it is one of the biggest drawbacks currently.

**Comment by eikaramba** [comment]
Dear bot this issue is still there. :)

**Comment by Telokis** [comment]
Any update on this?

**Comment by DanFerro** [comment]
Would love to know if there's a cleaner workaround.

**Comment by japhex** [comment]
Agreed, would really appreciate any word on this, also.

**Comment by GuillermoGaete** [comment]
+1 I have the same problem, 2 entites related by a JunctionEntity :(

**Comment by japhex** [comment]
I've seen the update saying this has been changed to have a bug status, can you guys confirm if that means you are actively looking to fix it? 

Would be amazing to know just I can plan ahead!

Thanks so much guys.

**Comment by deankostomaj** [comment]
Is there any news about this? I mean this is really basic stuff we are talking about and it is not yet handled even though latest version is 4.28. @mickhansen

**Comment by anonrig** [comment]
With the help of our dear friend Google, I've come to this issue. And sadly seen that it's not fixed yet.

**Comment by kauffmanes** [comment]
Any update for 2018?

**Comment by ashish277d** [comment]
Any updates on this issue?  Would love to know if there is any workaround?

**Comment by Bat-Orshikh** [comment]
unique: false still not working on belongsToMany association.

**Comment by akhoury** [comment]
2019?

**Comment by asgharmusani** [comment]
My problem was fixed with using `distinct:true`

**Comment by sliterok** [comment]
I think that issue represents the main point, why sequelize is awful orm.
If it wasn't an issue, than we could use db schema like this.
![image](https://user-images.githubusercontent.com/12751644/51408263-f4996180-1b7f-11e9-8661-d694be95e631.png)
But instead, sequelize urge you to use db schema like this instead. Which has duplicating values in it for no particular reason.
![image](https://user-images.githubusercontent.com/12751644/51408298-0d097c00-1b80-11e9-9d62-32bd4cab840e.png)

**Comment by Mayconbenito** [comment]
Any update on this?

**Comment by sprietNathanael** [comment]
More than 4 years and still no news. I think this is related to #5077. I'm looking for the best workaround for my case: I have two tables `A` and `B` that can be linked multiple times at different periods. I defined `AB` table with an `id` as primary key (`primarykey`, `unique`, `autoincrement`) and with `dateFrom` and `dateTo` fields.
Is there a way to add `dateFrom` and `dateTo` to the `unique` constraint between `Aid` and `Bid`?

**Comment by lfernando-silva** [comment]
"Solved" this performing an separated query to the N:M table (a raw query).

**Comment by papb** [comment]
I am under the impression that this can already be done in a clean way in Sequelize. I will make some tests later. If not, I will take this.

**Comment by sliterok** [comment]
Oof some progress here hurray!

**Comment by tobsfit** [comment]
Any news?

**Comment by EdilsonAndrade** [comment]
Hello, I am at the same page and no solution??? we are in 2020 already

**Comment by sliterok** [comment]
Yeah, that's why I'm using mongoose

**Comment by EdilsonAndrade** [comment]
> Yeah, that's why I'm using mongoose

I think that I will need to create a manually many to many :-1:

**Comment by Parakrama123** [comment]
still no solution?

**Comment by jclee14** [comment]
I just face this problem today, and wow! it still does not get fixed.

**Comment by sliterok** [comment]
> I just face this problem today, and wow! it still does not get fixed.

Yeah, there's 11 pages of open issues that have type: bug

**Comment by megasyl** [comment]
still no solution ?

**Comment by sliterok** [comment]
> still no solution ?

I guess the best solution is to use mongo + mongoose :(

**Comment by ShGKme** [comment]
> > still no solution ?
> 
> I guess the best solution is to use mongo + mongoose :(

@sliterok That's absolutely wrong. 

**Choosing DBMS model is not about features of some ORM.**

It's about **data model**, about problem specifics, about technical specifics. ORM is just abstraction layer and helpful library, that somethims helps you to work with relation DB.

So, about the issue and solution...

If you need M:M with duplicates and complex join table, just think about it as about two 1:M relations.

Instead of `A belongsToMany B throw C` you have `A hasMany C` and `C belongsTo B`. It gives you full control (with a bit more complex queries).

**Comment by Abdulrahman-Falyoun** [comment]
Still no solution?

**Comment by 7hibault** [comment]
Also having this issue in sept 2020

**Comment by imjordanxd** [comment]
Still having this in October 2020

**Comment by eperezf** [comment]
Nov. 2020, issue is still there.

**Comment by Jonny137** [comment]
I hope Santa will find a solution by January.

**Comment by imjordanxd** [comment]
@eperezf @Jonny137 I don't expect this issue to ever get fixed. A _decent_ workaround is to implement using one-to-many relationships.

**Comment by pepsivontwist** [comment]
@imjordanxd seems like with sequelize anything you need to do requires a workaround these days; guess it's time to move on

**Comment by SheaOHagan** [comment]
Also having this problem, would be keen to hear about a solution when available.

**Comment by b-karin** [comment]
Same problem

**Comment by festo** [comment]
OMG I cannot believe this is still not resolved

**Comment by dr-miracle** [comment]
Find some [workaround](https://gist.github.com/mx781/dc6b86fad71c394872834397c3f8a52f). 
It's a little bit complicated to read and similar to what was suggested @ShGKme. I guess this is the only way to do it.

**Comment by matthieu-alc** [comment]
The only solution i found is to define the 3 models and the 3 parts of the association like I explained here :
https://stackoverflow.com/a/68513372/16503781

**Comment by pepsivontwist** [comment]
@matthieu-alc Yeah, but then every time you query it you have to define the whole nested include.
6 years and counting on this issue, just saying...

**Comment by tresloukadu** [comment]
2021 Aug still counting

**Comment by GuillermoGaete** [comment]
> 2021 Aug still counting

It never will be added/fixed.

**Comment by MustafaJamal** [comment]
Will it be implemented? Please make it work :( TIA

**Comment by imjordanxd** [comment]
I've come to accept that it'll never be fixed.

**Comment by tresloukadu** [comment]
got another situation and now I have to make 2 queries..... why the developers ignore this problem ?

**Comment by tresloukadu** [comment]
Now is too late to change this lib in my project so I had to accept  the possibility of doing 2 queries.

For newer projects I'm will using KNEX.

**Comment by gu1lhem** [comment]
After a lot of research I have finnaly been brought here...

**Comment by niraj-khatiwada** [comment]
Encountered this issue today. Surprise to see its not fixed yet even after 6 years.

**Comment by matthieu-alc** [comment]
> Encountered this issue today. Surprise to see its not fixed yet even after 6 years.
> 
> EDIT: This solution works for me: https://stackoverflow.com/a/58439584/11806970

This solution should work, make sure you are setting the unique property to false this way :

`User.belongsToMany(Item, { through: { model: UserItem, unique: false } });`

**Comment by selvesandev** [comment]
Jan 18, 2022 

![image](https://user-images.githubusercontent.com/21096850/149943773-1fb83ce4-04a9-4bfc-a4a1-cbea7a9a2b27.png)

**Comment by WikiRik** [comment]
Nope, feel free to submit a PR that fixes this issue. We're happy to help with that if you need assistance when working on it

**Comment by SkoobGeneral** [comment]
I just ran into this bug and added this options to the main query:

```
{
  raw: true,
  plain: false, 
  nest: true
}

```
Then you just merge the query.

It's a workaround, but might help somebody.

**Comment by SkoobGeneral** [comment]
Many-to-many "through" table with multiple rows of identical foreign key pairs only returns one result?

I just ran into this bug and added this options to the main query:

```
    {
       raw: true,
       plain: false, 
       nest: true
     }
```

Then you just merge the query.

It's a workaround, but might help somebody.

**Comment by fjobeir** [comment]
I am using **sequelize 6.28** and surprisingly, this is how you solve it:
Assuming the model Foo has a **many to many** relationship with model Bar, so in Foo.js we have:
`Foo.belongsToMany(Bar, {
        through: 'Views',
        foreignKey: 'fooId',
        as: 'View'
      })`

and in Bar.js we have
`Bar.belongsToMany(Foo, {
        through: 'Views',
        foreignKey: 'barId',
        as: 'View'
      })`

Referring to the mixin/special methods, we can add a relation between those 2 models be using this function:
`Foo.addBar(barId)`

This code will create a relationship but it will fail when you try to add the relationship again. All what you have to do is to **convert the _barId_ to a string**.
`Foo.addBar(String(barId))`

I should dive into Sequelize code and follow up what is going on there

**Comment by iradofurioso** [comment]
> I am using **sequelize 6.28** and surprisingly, this is how you solve it: Assuming the model Foo has a **many to many** relationship with model Bar, so probably we have something like that:
> 
> `Foo.belongsToMany(Bar, { through: 'Views', foreignKey: 'fooId', as: 'View' })`
> 
> and
> 
> `Bar.belongsToMany(Foo, { through: 'Views', foreignKey: 'barId', as: 'View' })`
> 
> Referring to the mixin/special methods, we can add a relation between those 2 models by using this function:
> 
> `Foo.addBar(barId)`
> 
> This code will create a relationship but it will fail when you try to add the relationship again. All what you have to do is to **convert the _barId_ to a string**.
> 
> `Foo.addBar(String(barId))`
> 
> I should dive into Sequelize code and follow up what is going on there

well as it seems... this problem was not solved yet!

**Comment by Nef007** [comment]
I inform you that the problem is still relevant, I did not find a worthy solution unique: false does not work!

version 6.31.1

**Comment by The-Lone-Druid** [comment]
Has anyone found any solutions guys? it's like 2024 now 🥺

**Comment by sliterok** [comment]
At this point this issue should be disclaimer when you first discover sequelize. I just checked [many-to-many example](https://playground.prisma.io/guides/relations_many-to-many-explicit?step=2) on prisma playground and it seems to work fine with schema similar to the example in the first post.

**Comment by fjobeir** [comment]
@The-Lone-Druid @sliterok 
Have you guys tried what I commented?
https://github.com/sequelize/sequelize/issues/3493#issuecomment-1382716740

**Comment by The-Lone-Druid** [comment]
> @The-Lone-Druid @sliterok Have you guys tried what I commented? [#3493 (comment)](https://github.com/sequelize/sequelize/issues/3493#issuecomment-1382716740)

Thanks @fjobeir I did try, but it didn't seem to solve the issue that I'm facing, I had to migrate my ORM to Prisma and thankfully it resolved the issue.
